File: 4kyu/Explosive Sum/main.py
from functools import lru_cache


@lru_cache(maxsize=None)
def exp_sum(n: int) -> int:
    parts = 0

    if n in [0, 1]:
        return 1
    else:
        for i in range(1, n + 1):
            p1 = n - int((i * (3 * i - 1)) / 2)
            p2 = n - int((i * (3 * i + 1)) / 2)
            sign = pow(-1, i + 1)
            parts += sign * (exp_sum(p1) + exp_sum(p2))

        return parts

File: 4kyu/Explosive Sum/main.cpp
using ull = unsigned long long;

ull exp_sum(unsigned int n) {
    ull dp[n+1];
    dp[0] = 1;

    for (unsigned int i = 1; i <= n; i++) {
        dp[i] = 0;
    }

    for (unsigned int k = 1; k <= n; k++) {
        for (unsigned int i = k; i <= n; i++) {
        dp[i] += dp[i-k];
        }
    }

    return dp[n];
}